WordPress/gutenberg
# ColorPalette, BorderControl, GradientPicker: refine types and logic around single vs multiple palettes

## What?


As a follow-up to #46315, this PR further cleans up some of the logic around single vs multiple origin color palettes.

## Why?


Code quality (re-use same checks in different components, add comments, tighten up types)

## How?


Here's the main set of changes applies in this PR:

- moved `ColorPalette` utilities to a separate `utils` file
- rewrote slightly the "multiple origins" check for added clarity (and added the array `lenght` check)
- removed a check in `ColorPalette` that isn't necessary anymore since the removal of the `__experimentalHasMultipleOrigins` prop in #46315
- reuse the `ColorPalette` logic in `BorderControl` instead of custom check (which allows to simplify some types and checks too!)
- add similar logic for `GradientPicker` too

You can review this PR commit-by-commit too

## Testing Instructions





In Storybook, check that `ColorPalette`, `GradientPicker` and `BorderControl` components behave like on `trunk`. Pay particular attention to testing these components with single vs multiple origin colors/gradients.

Smoke test the block editor and make sure that the above components do not show any regressions.
